src/Analyzers/CSharp/Analyzers/NamespaceFileSync/NamespaceFileSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.LanguageServices;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.NamespaceFileSync
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class NamespaceFileSyncDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private static readonly LocalizableResourceString s_localizableTitle = new LocalizableResourceString(
+          nameof(CSharpAnalyzersResources.Namespace_named_incorrectly), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        private static readonly LocalizableResourceString s_localizableInsideMessage = new LocalizableResourceString(
+            nameof(CSharpAnalyzersResources.Namespace_must_match_folder_structure), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources));
+
+        public const string RootNamespaceOption = "build_property.RootNamespace";
```

Comment:
Is there a way to make sure this is in-sync with the target files?

Reply:
I can add to a centralized location. They're always going to be "build_property.<Name>"

Reply:
oh nm, this is only place it's used and target files will always use well known msbuild names. I don't know if this could be shared in the way you're thinking, but since it's generated and based on known properties already it will be fine.